infra/main.tf

```diff
@@ -435,293 +435,102 @@ resource "kubernetes_service" "ukbb_rb_static" {
-  }
-}
-
-resource "random_id" "test_name_suffix" {
-  byte_length = 2
-}
-
-resource "google_service_account" "test" {
-  account_id = "test-${random_id.test_name_suffix.hex}"
-}
-
-resource "google_service_account_key" "test_key" {
-  service_account_id = google_service_account.test.name
-}
-
-resource "kubernetes_secret" "test_gsa_key" {
-  metadata {
-    name = "test-gsa-key"
-  }
-
-  data = {
-    "key.json" = base64decode(google_service_account_key.test_key.private_key)
-  }
-}
-
-resource "google_project_iam_member" "test_compute_instance_admin" {
-  role = "roles/compute.instanceAdmin.v1"
-  member = "serviceAccount:${google_service_account.test.email}"
-}
-
-resource "google_project_iam_member" "test_service_account_user" {
-  role = "roles/iam.serviceAccountUser"
-  member = "serviceAccount:${google_service_account.test.email}"
-}
-
-resource "google_project_iam_member" "test_logging_viewer" {
-  role = "roles/logging.viewer"
-  member = "serviceAccount:${google_service_account.test.email}"
-}
-
-resource "google_project_iam_member" "test_service_usage_consumer" {
-  role = "roles/serviceusage.serviceUsageConsumer"
-  member = "serviceAccount:${google_service_account.test.email}"
-}
-
-resource "google_service_account" "test_dev" {
-  account_id = "test-dev"
-}
-
-resource "google_service_account_key" "test_dev_key" {
-  service_account_id = google_service_account.test_dev.name
-}
-
-resource "kubernetes_secret" "test_dev_gsa_key" {
-  metadata {
-    name = "test-dev-gsa-key"
-  }
-
-  data = {
-    "key.json" = base64decode(google_service_account_key.test_dev_key.private_key)
-  }
+module "test_dev_gsa_secret" {
+  source = "./gsa_k8s_secret"
+  name = "test-dev"
 }
 
 resource "google_service_account" "batch_agent" {
   account_id = "batch2-agent"
 }
 
-resource "google_project_iam_member" "batch_agent_compute_instance_admin" {
-  role = "roles/compute.instanceAdmin.v1"
-  member = "serviceAccount:${google_service_account.batch_agent.email}"
-}
-
-resource "google_project_iam_member" "batch_agent_service_account_user" {
-  role = "roles/iam.serviceAccountUser"
-  member = "serviceAccount:${google_service_account.batch_agent.email}"
-}
-
-resource "google_project_iam_member" "batch_agent_log_writer" {
-  role = "roles/logging.logWriter"
-  member = "serviceAccount:${google_service_account.batch_agent.email}"
-}
-
-resource "google_project_iam_member" "batch_agent_object_creator" {
-  role = "roles/storage.objectCreator"
-  member = "serviceAccount:${google_service_account.batch_agent.email}"
-}
+resource "google_project_iam_member" "batch_agent_iam_member" {
+  for_each = toset([
+    "compute.instanceAdmin.v1",
+    "iam.serviceAccountUser",
+    "logging.logWriter",
+    "storage.objectCreator",
+    "storage.objectViewer",
+  ])
 
-resource "google_project_iam_member" "batch_agent_object_viewer" {
-  role = "roles/storage.objectViewer"
+  role = "roles/${each.key}"
```

Comment:
Just double checking this shouldn't be `for_each` here.

Reply:
Ya, this is a weird quirk of the terraform language (HCL). The grammar is very strict, basically just blocks with `key = val` pairs, which doesn't exactly lend itself to writing for loops. The way they get around that is by saying that if you assign the reserved keyword `for_each` to a set or map in a resource block, it will then create a resource for each element in the assigned set, and assign that value to the reserved binding `each`.